### PR TITLE
Fixed accesses for annual reports

### PR DESCRIPTION
**Pull request by VSBalanchuk**
dev
## JIRA

* [Main JIRA ticket](https://jira.softserve.academy/secure/RapidBoard.jspa?rapidView=id)

## Code reviewers

- [ ] @github_username

### Second Level Review

- [ ] @github_username

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: EPlast/EPlast.BLL/Services/AnnualReport/AnnualReportAccessService.cs [hunk 1]
```diff
@@ -13,11 +15,50 @@ public AnnualReportAccessService(IRepositoryWrapper repositoryWrapper)
...
+                        Roles.AdminAndOkrugaHeadAndOkrugaHeadDeputy.Contains(x.AdminType.AdminTypeName))
+                    : null;
+                return regionAnnualReportRegionAdministration != null;
+            }
+
+            return false;
+        }
 
-        public async Task<bool> CanEditCityReportAsync(string userId, int reportId)
+        private enum ReportType
```
Comment (andrii52) [inline EPlast/EPlast.BLL/Services/AnnualReport/AnnualReportAccessService.cs:57]: Move this to Rresources project

Comment (VSBalanchuk) [reply EPlast/EPlast.BLL/Services/AnnualReport/AnnualReportAccessService.cs]: Thanks, changed it

**Review by Bodasyuk — APPROVED** [review]

File: EPlast/EPlast.BLL/Services/AnnualReport/AnnualReportAccessService.cs [hunk 2]
```diff
@@ -13,11 +15,43 @@ public AnnualReportAccessService(IRepositoryWrapper repositoryWrapper)
...
+                    await _repositoryWrapper.ClubAnnualReports.GetFirstOrDefaultAsync(x => x.ID == reportId);
+                var clubAnnualReportClubAdministration = clubAnnualReport != null
+                    ? await _repositoryWrapper.ClubAdministration.GetFirstOrDefaultAsync(x =>
+                        x.ClubId == clubAnnualReport.ClubId && x.UserId == user.Id &&
+                        x.AdminType.AdminTypeName != null &&
+                        Roles.AdminAndKurinHeadAndKurinHeadDeputy.Contains(x.AdminType.AdminTypeName))
+                    : null;
+                return clubAnnualReportClubAdministration != null;
+            }
+            else if (myReportType == ReportType.Region)
```
Comment (OleksandrBilan) [inline EPlast/EPlast.BLL/Services/AnnualReport/AnnualReportAccessService.cs:41]: there is 'switch' statement invented for such purposes

**Review by IlliaDR — APPROVED** [review]

File: EPlast/EPlast.BLL/Interfaces/UserAccess/IUserAccessService.cs [hunk 3]
```diff
@@ -30,9 +30,13 @@ public interface IUserAccessService
...
         /// Returns dictionary with user accesses for annual reports
-        /// cityReportId is optional if we want to check user access for certain city report
+        /// report type and report id is optional
+        /// report type
+        /// 0 - city
+        /// 1 - club
+        /// 2 - region
         /// </summary>
-        Task<Dictionary<string, bool>> GetUserAnnualReportAccessAsync(string userId, int? cityReportId = null);
+        Task<Dictionary<string, bool>> GetUserAnnualReportAccessAsync(string userId, User user, int? reportType = null, int? reportId = null);
```
[the remark this answers is not in the document]
Comment (VSBalanchuk) [reply EPlast/EPlast.BLL/Interfaces/UserAccess/IUserAccessService.cs]: Thanks, changed it

File: EPlast/EPlast.BLL/Services/UserAccess/UserAccessService.cs [hunk 4]
```diff
@@ -94,13 +94,13 @@ public UserAccessService(IClubAccessService clubAccessService, IEventUserAccessS
...
-        public async Task<Dictionary<string, bool>> GetUserAnnualReportAccessAsync(string userId, int? cityReportId=null)
+        public async Task<Dictionary<string, bool>> GetUserAnnualReportAccessAsync(string userId, User user, int? reportType=null, int? reportId=null)
         {
             _securityModel.SetSettingsFile(AnnualReportSecuritySettingsFile);
             var userAccess = await _securityModel.GetUserAccessAsync(userId);
-            if (cityReportId != null)
+            if (reportType!= null && reportId != null)
             {
-                userAccess["EditReport"] = await _annualReportAccessService.CanEditCityReportAsync(userId, (int)cityReportId);
+                userAccess["EditReport"] = await _annualReportAccessService.CanEditReportAsync(user, (int)reportType, (int)reportId);
```
[the remark this answers is not in the document]
Comment (VSBalanchuk) [reply EPlast/EPlast.BLL/Services/UserAccess/UserAccessService.cs]: Thanks, changed it

File: EPlast/EPlast.BLL/Services/AnnualReport/AnnualReportAccessService.cs (hunk 2 above)
Comment (VSBalanchuk) [reply EPlast/EPlast.BLL/Services/AnnualReport/AnnualReportAccessService.cs]: Thanks, changed it

File: EPlast/EPlast.BLL/Services/AnnualReport/AnnualReportAccessService.cs [hunk 5]
```diff
@@ -13,11 +15,43 @@ public AnnualReportAccessService(IRepositoryWrapper repositoryWrapper)
             _repositoryWrapper = repositoryWrapper;
         }
 
-
-        public async Task<bool> CanEditCityReportAsync(string userId, int reportId)
+        public async Task<bool> CanEditReportAsync(User user, int reportType, int reportId)
         {
-            var annualReport = await _repositoryWrapper.AnnualReports.GetFirstOrDefaultAsync(x => (x.CreatorId == userId && x.ID == reportId));
-            return annualReport != null;
+            var myReportType = (ReportType) reportType;
```
[the remark this answers is not in the document]
Comment (VSBalanchuk) [reply EPlast/EPlast.BLL/Services/AnnualReport/AnnualReportAccessService.cs]: Thanks, changed it

**Review by Liubov-H — APPROVED** [review]

File: EPlast/EPlast.Resources/AnnualReportType.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+﻿namespace EPlast.Resources
+{
+    public enum ReportType
+    {
+        City = 0,
```
Comment (ostappppppp) [inline EPlast/EPlast.Resources/AnnualReportType.cs:5]: You don't need to specify numbers here, it starts with 0 by default

Comment (VSBalanchuk) [reply EPlast/EPlast.Resources/AnnualReportType.cs:5]: Thanks, changed it

**Pull request merged**